### PR TITLE
Support ufunc call in CuPy JIT

### PR DESCRIPTION
**Pull request by asi1024**
Part of #4290.

### CODE REVIEW

**Comment by emcastillo** [comment]
I am a bit concerned with the ufunc being a `__global__` function in the case of most of the declared ufuncs in cupy.
Although in CuPy `__calling__` a global func, from inside another `__global__` func is working. It is technically not allowed in CUDA. We should resort to the `<<<...>>>` syntax. @leofang any thoughts here?

**Comment by asi1024** [comment]
This PR supports `ufunc` only for scalar operation.

CuPy JIT is currently used only from `cupy.vectorize`, and we don't need to consider supporting for ndarray input. Also, I have no plan for ndarray input support in the future also for CuPy JIT raw kernel definition.

**Comment by leofang** [comment]
Not sure if I got the gist here. Does this PR allow the JIT to emit something equivalent of this?
```C++
template<typename T>
__global__ void my_ufunc(const T& a, T& b) {
    b += a;
}

__global__ void my_jit_ker(const float* a, float* b, unsigned int N) {
    unsigned int tid = ...
    if (tid < N) {
        my_ufunc(a[tid], b[tid]);
    }
}
```
If so, the concern would be the call to `my_ufunc` is illegal as it lacks of the launch syntax `<<< >>>`. NVRTC happens to allow this and it seems to work fine, but both NVCC and HIPRTC will fail to compile it.

**Comment by asi1024** [comment]
@leofang No, ufunc is represented with device function!

**Comment by asi1024** [comment]
Now ufunc call is supported.
In this PR, ufunc call is represented with device function in the generic case, but in special cases, inline expansion is enabled for the readability of generated CUDA code.

For example, `cupy.add` uses inline expansion.
```py
__device__ double f(double x, double y) {
  return (x + y);
}
```

On the other hand, `cupy.remainder` does not use inline expansion because `in0` occurs twice.
```py
template <typename T0, typename T1>
__device__ double cupy_remainder_float64(T0 in0, T1 in1) {
    return in0 - _floor_divide(in0, in1) * in1;
}

__device__ double f(double x, double y) {
  return cupy_remainder_float64(x, y);
}
```

**Comment by asi1024** [comment]
PTAL.

File: cupyx/jit/_compile.py [hunk 1]
```diff
@@ -172,19 +191,43 @@ def _eval_operand(op, args, env, dtype=None):
...
+                expr = expr.replace(f'in{i}', x.code)
+            expr = '(' + expr.replace('out0_type', str(out_type)) + ')'
+            env.preambles.add(ufunc._preamble)
+        else:
+            template_typenames = ', '.join([
+                f'typename T{i}' for i in range(ufunc.nin)])
+            ufunc_name = f'{ufunc.name}_{str(numpy.dtype(op.out_types[0]))}'
+            params = ', '.join([f'T{i} in{i}' for i in range(ufunc.nin)])
+            ufunc_code = f"""template <{template_typenames}>
+__device__ {out_type} {ufunc_name}({params}) {{
```
Comment (leofang) [inline cupyx/jit/_compile.py:221]: Do we wanna inline it?
```suggestion
__device__ __forceinline__ {out_type} {ufunc_name}({params}) {{
```

File: cupyx/jit/_compile.py [hunk 2]
```diff
@@ -346,9 +401,23 @@ def _transpile_expr(expr, env):
...
+}
+
+
+def _astype_scalar(x, ctype, unsafe=False):
+    from_t = x.ctype.dtype
+    to_t = ctype.dtype
+    if from_t == to_t:
         return x
+    if not unsafe:
+        if _scalar_kind_score[from_t.kind] > _scalar_kind_score[to_t.kind]:
```
Comment (leofang) [inline cupyx/jit/_compile.py:419]: Would it be better to just rely on `cp.can_cast` to avoid corner cases?

File: cupyx/jit/_compile.py (hunk 1 above)
Comment (asi1024) [reply cupyx/jit/_compile.py:221]: The inline expansion I commented on https://github.com/cupy/cupy/pull/4347#issuecomment-766612103 is for the readability of generated code, not for the performance improvement. Do you think `__forceinline__` is needed for the performance?

Comment (emcastillo) [reply cupyx/jit/_compile.py:221]: I wonder the real effect of this.

Comment (leofang) [reply cupyx/jit/_compile.py:221]: Let's go without it and trust the compiler's heuristics! 🙂

**Comment by emcastillo** [comment]
Jenkins, test this please

**Review by emcastillo — APPROVED** [review]

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by asi1024** [comment]
Jenkins, test this please.

**Review by emcastillo — APPROVED** [review]

**Comment by emcastillo** [comment]
Seems like some very light precision issues

**Comment by asi1024** [comment]
Jenkins, test this please.

**Pull request merged**